Implement "update the image data"

gterzian opened the pull request:
<!-- Please describe your changes on the following line: -->
Spec compliant implementation of the [update the image data algorithm](https://html.spec.whatwg.org/multipage/embedded-content.html#update-the-image-data).

Currently still a work in progress, the ['async src complete test`](https://github.com/servo/servo/blob/master/tests/wpt/web-platform-tests/html/semantics/embedded-content/the-img-element/img.complete.html#L33) is still passing as it was before, even though I switched to the new code, so I guess that's something. 

@jdm I will be picking this up next weekend, I left a bunch of TODO and NOTES in the code, if you or someone else have time this week I would appreciate an initial scan and feedback. 

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [ ] `./mach build -d` does not report any errors
- [ ] `./mach test-tidy` does not report any errors
- [ ] These changes fix #11517 (github issue number if applicable).

<!-- Either: -->
- [x] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/16238)
<!-- Reviewable:end -->

gterzian committed:
improve spec compliance of update the image data

highfive commented:
Heads up! This PR modifies the following files:
 * @fitzgen: components/script/dom/htmlimageelement.rs
 * @kichjang: components/script/dom/htmlimageelement.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify script code, but no tests are modified. Please consider adding a test!

jdm commented:
Wonderful! Thanks for tackling this!

jdm commented:
Note: there's no need to include any implementation related to srcset/sizes/other responsive image things, as that's going to be a student project starting this summer.

gterzian commented:
@jdm great, glad to hear that! That would include concepts like [selecting and image source](https://html.spec.whatwg.org/multipage/embedded-content.html#selecting-an-image-source) and [updating the source set](https://html.spec.whatwg.org/multipage/embedded-content.html#updating-the-source-set), correct?

jdm commented:
Yep, that sounds great.

emilio commented:
r? @jdm

gterzian commented:
@jdm here is this weekend's harvest:

1. All [`img.complete` tests](https://github.com/servo/servo/blob/master/tests/wpt/web-platform-tests/html/semantics/embedded-content/the-img-element/img.complete.html) are now passing.
2. I had to edit one test, because from what I could tell the `while` loop that was in there was preventing the page from reaching ths `stable state`, therefore the test was always bound to fail. 
3. I tried to get the fetch part of the code to be a bit more spec compliant, meaning following the comcept of the `image request`, as well as adding some things related to step 14 of the spec. 
4. I made `Complete` more spec compliant. 

Re 3, the previous code would always just use `self.current_request`, however to follow the spec we must first pick the `image_request` from either the pending or current one, and then associate the fetch code with this `image_request`. Due to the fact that `ImageRequest` isn't thread safe in it's current form, I couldn't pass the actual request along the algorithm, so instead I pass around the url and grab the 'right' request based on the url when needed. I "think" that's pretty good for now...

Re 4, I had to decide how to have `Complete` return true in the case that "The final task that is queued by the networking task source once the resource has been fetched has been queued". I didn't want to add more state to `ImageRequest`, so instead I made the choice that if `current_request` has a parsed url, but isn't yet in either `Broken` or `CompletelyAvailable` state, we can assume the final task has been queued but not processed yet... Once again, I "think" that's good enough for now... 

If you have time this week, I think the PR is good for a more comprehensive review at this stage...

gterzian commented:
@jdm I'm reading through the spec again and I see I missed some key concepts regarding current and pending requests in step 12 and 14 of the specs, so that's certainly something I will still need to implement, feedback in the meantime welcome off course...

jdm commented:
Re: "The final task queued that is queued", see https://github.com/whatwg/html/issues/1055.

gterzian commented:
@jdm thanks, so that means we can keep that test as it was before, and with a `FAIL`? Glad to read that because it wasn't entirely clear to me. 

Re pending/current request, here is the work I intend to still do, next weekend:

1. make `pending_requent` an `Option<DOMRefCell<ImageRequest>>`, and initially set it to `None`. 
2. Make the fetch and 'choose a request' code comply with step 12 and 14 of the spec, for example the whole 'abort current and upgrade pending to current`. Also currently I just `return` in some cases, while in fact the spec speaks of `aborting these steps`, which I guess means just aborting step 12, not returning form the whole algorithm(This relates to picking the right request to fetch the image).

So it's mainly looking at step 12 and 14 in more details....

jdm commented on components/script/dom/htmlimageelement.rs:
```diff
@@ -271,101 +264,319 @@ impl HTMLImageElement {
-        self.current_request.borrow_mut().metadata = metadata;
 
         // Mark the node dirty
         self.upcast::<Node>().dirty(NodeDamage::OtherNodeDamage);
```
We should only do this if the current request is the one being serviced.

jdm commented on tests/wpt/web-platform-tests/html/semantics/embedded-content/the-img-element/img.complete.html:
```diff
@@ -33,16 +33,20 @@
-    document.getElementById("imgTestTag3").src = '3.jpg?nocache=' + Math.random();
+    document.getElementById("imgTestTag3").onload = t.step_func(function(){
+      assert_true(document.getElementById("imgTestTag3").complete);
+      assert_equals(document.getElementById("imgTestTag3").src.split("/").pop().split('?')[0], "3.jpg");
```
What if we use currentSrc here instead of src?

jdm commented on tests/wpt/web-platform-tests/html/semantics/embedded-content/the-img-element/img.complete.html:
```diff
@@ -51,14 +55,16 @@
-      if (imageInstance.complete === true) break;
-    }
-    t.done();
+    setTimeout(function(){
```
`t.step_timeout(function {`

jdm commented on components/script/dom/htmlimageelement.rs:
```diff
@@ -568,8 +779,26 @@ impl HTMLImageElementMethods for HTMLImageElement {
+                // NOTE: this is an attempt to comply with:
+                // "(return true if)The final task that is queued by the networking task source,
+                //  once the resource has been fetched, has been queued"
+                return request.parsed_url.is_some()
```
Let's forget about this.

jdm reviewed (changes requested):
Great work so far! I found these changes to be quite readable in general, although it might be nice to follow the spec text about aborting a particular request or creating a new request, rather than what is (presumably) equivalent.

gterzian commented:
@jdm thanks for the review, I'll look at it in more details over the weekend...

gterzian commented on components/script/dom/htmlimageelement.rs:
```diff
@@ -113,7 +116,8 @@ impl Runnable for ImageResponseHandlerRunnable {
         if element.generation.get() == self.generation {
-            element.process_image_response(self.image);
+            let url = self.url.clone();
+            element.process_image_response(self.image, &url);
```
yes indeed, and I also had to use `&self.image` to prevent `use of moved value: self`

gterzian commented on components/script/dom/htmlimageelement.rs:
```diff
@@ -271,101 +264,319 @@ impl HTMLImageElement {
+    fn update_the_image_data_sync_steps(&self) {
+        let document = document_from_node(self);
+        // Step 8
+        let image_source = self.select_image_source();
```
On this one I have a question: have you considered that the spec mentions "selecting a source" as part of the sync steps, once a so-called 'stable state' has been reached, and that the sync steps will run only once, even if `update_the_image_data` was called several times. Since `update_the_image_data` should be called if the source changes, perhaps when we reach this point we in fact will always get the "latest" source?

gterzian commented on components/script/dom/htmlimageelement.rs:
```diff
@@ -271,101 +264,319 @@ impl HTMLImageElement {
+                                                                  UsePlaceholder::No,
+                                                                  CanRequestImages::No);
+                if let Ok(ImageOrMetadataAvailable::ImageAvailable(image)) = response {
+                    // Step 5.3
```
added it, see below

gterzian commented on components/script/dom/htmlimageelement.rs:
```diff
@@ -271,101 +264,319 @@ impl HTMLImageElement {
 
         // Fire image.onerror
         if trigger_image_error {
             self.upcast::<EventTarget>().fire_event(atom!("error"));
```
thanks, added it as well, see below

gterzian replied:
done below

gterzian commented:
@jdm OK I have addressed most of your comments, I just had a question re https://github.com/servo/servo/pull/16238/files/f3ff232ada8aa4038add19a5dbfcc9bdc32054fa#r110809997, and I have also made further attempt to implement the concepts of pending and current requests in a more spec compliant way.

gterzian commented on components/script/dom/htmlimageelement.rs:
```diff
@@ -271,101 +264,319 @@ impl HTMLImageElement {
+        image_request.source_url = Some(src.clone());
+        image_request.image = None;
+        image_request.metadata = None;
+        let document = document_from_node(self);
```
the actual fetch request is initialized a bit futher down the road, [here](https://github.com/servo/servo/blob/master/components/script/dom/htmlimageelement.rs#L261), after a cache check. I'm not sure if we need to change this, even it is indeed at this stage(step 12.6) that the spec speaks of "creating a potential-CORS request"...

gterzian commented:
@jdm it's ready for another round of review, if you have time this week. In particular, I would like to get feedback on:

1. The new `ImageRequestPhase` concept I've introduced. The spec mentions the concept of the 'image request' with whom the fetch part of the algorithm should be associated. This 'image request' should either be the 'current' or 'pending' request, and there is quite a bit of logic associated with choosing either and "upgrading" the "image request" from the one or the other. 
I figured a good way to implement this logic is by adding a `image_request` field, which is either `ImageRequestPhase::Pending` or `ImageRequestPhase::Current`. In parallel the `current_request` and `pending_request` need to be mutated as well as per the logic of the algorithm, however we still need to know which one is the "image request"...
2. I'm wondering if my changes to the wpt tests should be kept or discarded, since we've concluded that the one test related to the "final task" logic should be left as `FAIL`, and my changes were mostly an attempt to get it to pass. 

Other than that, I've had another good look at Step 12 and 14, where most of the current/pending request stuff is found, and I *think* that it's all mostly there. Off course it's hard to tell for sure since there are no tests, and I don't think we can really test it with wpt tests, around that concept.

gterzian commented:
@jdm by the way, one thing that is missing is: "That task, and each subsequent task, that is queued by the networking task source while the image is being fetched, if image request is the current request, must update the presentation of the image appropriately (e.g., if the image is a progressive JPEG, each packet can improve the resolution of the image)."

In fact I have left most of the previous fetch/cache check code intact, and tried to implement the various parts of Step 14 in `process_image_response`. 

That function is always called as a 'cache listener', and it's the image cache that processes the actual network responses. 

From what I can tell from the image cache implementation, it seems that currently in case of new chunk of image data coming in, it is is only the image metadata that is 'updated' and returned as `ImageResponse::MetadataLoaded(meta)`(see https://github.com/servo/servo/blob/master/components/net/image_cache.rs#L500), and `process_image_response` will then update the metadata of the "image request". 

I guess the API of the image cache and it's implementation could be adapted to also return some appropriate intermediary image, for example in the case of  progressive JPEG... 

I also guess in the meantime we can consider that "update the presentation of the image appropriately" could in our case mean, "update the metadata", and then only update the actual image as part of the "last task that is queued by the networking task source", thereby skipping any potential intermediary "update the presentation of the image". 

In other words, we're currently just updating the metadata in case of an incoming chunk, and only update the actual image when it has been completely fetched...

jdm commented on components/script/dom/htmlimageelement.rs:
```diff
@@ -174,18 +182,7 @@ impl PreInvoke for ImageContext {}
-                Some(LoadBlocker::new(&*document, LoadType::Image(img_url.clone())));
-        }
-
+    fn fetch_image(&self, img_url: &ServoUrl) {
```
Let's make this method accept a `&mut Option<LoadBlocker>` and create it here. This will match the specification: "Fetching the image must delay the load event of the element's node document"

gterzian replied:
There is a comment a few lines below that reads:
```
// This is a background load because the load blocker already fulfills the
// purpose of delaying the document's load event.
```
Since we are indeed creating and terminating load blockers as part of the "image request" cycle, do we need to create one more blocker here?

jdm replied:
That's a fair point.

jdm commented on components/script/dom/htmlimageelement.rs:
```diff
@@ -271,106 +268,360 @@ impl HTMLImageElement {
-        self.generation.set(self.generation.get() + 1);
+    /// https://html.spec.whatwg.org/multipage/#abort-the-image-request
+    fn abort_request(&self, state: State, request: ImageRequestPhase) {
+        match request {
```
Let's have this method terminate the appropriate load blocker for the given request.

gterzian replied:
is this not dealt with through `LoadBlocker::terminate(&mut request.blocker);`?

jdm commented on components/script/dom/htmlimageelement.rs:
```diff
@@ -271,106 +268,360 @@ impl HTMLImageElement {
+                    width: image.width
+                });
+                self.current_request.borrow_mut().state = State::CompletelyAvailable;
+                LoadBlocker::terminate(&mut self.current_request.borrow_mut().blocker);
```
The danger here of only operating on the current request is that the load blocker for the pending request will remain unterminated.

gterzian replied:
I switched to first terminating the pending requent, then switching to using the current request. I think previously I was following the spec text too closely, which didn't match how we implemented the concept of pending/current requests...

jdm commented on components/script/dom/htmlimageelement.rs:
```diff
@@ -271,106 +268,360 @@ impl HTMLImageElement {
+        }
+    }
 
+    fn init_pending_request(&self,
```
Why not make this a free `init_image_request(&mut ImageRequest, url: &ServoUrl, src: &DOMString)` function so we can use it for the current request as well?

gterzian replied:
Good idea, although I implemented slightly differently because of the need to access `self`, and also because it wasn't so easy to pass around the `RefMut<ImageRequest>`...

gterzian replied:
EDIT: had to pass the request around in the end, to avoid multiple borrowing of it...

jdm commented on tests/wpt/web-platform-tests/html/semantics/embedded-content/the-img-element/img.complete.html:
```diff
@@ -33,16 +33,20 @@
-    document.getElementById("imgTestTag3").src = '3.jpg?nocache=' + Math.random();
+    document.getElementById("imgTestTag3").onload = t.step_func(function(){
+      assert_true(document.getElementById("imgTestTag3").complete);
+      assert_equals(document.getElementById("imgTestTag3").currentSrc.split("/").pop().split('?')[0], "3.jpg");
```
What if we do `new URL(document.getElementById('imgTestTag3').currentSrc).pathname` instead of mucking with the string?

jdm replied:
Actually, why don't we verify that we only see the second src value that was set?

jdm commented on tests/wpt/web-platform-tests/html/semantics/embedded-content/the-img-element/img.complete.html:
```diff
@@ -51,14 +55,16 @@
+      assert_false(loaded, "image should not have loaded yet");
+      assert_true(imageInstance.complete, "imageInstance.complete should be true");
+      t.done();
+    }, 2000);
```
Let's revert the changes to this test. I'd prefer to make changes like this upstream since this part of the spec is a bit contentious.

jdm reviewed (changes requested):
Sorry for not getting to this sooner! I'm really pleased with these changes - with the exception of the ImageRequestPhase model, it was straightforward to match the implementation against the specification text. I was going to argue against ImageRequestPhase, but I haven't come up with a better representation of what the specification describes.

I've proposed some changes which should reduce some of the code duplication, as well as pointed out a couple places where we still don't match the spec in terms of async tasks. I think this is pretty close!

gterzian commented:
@jdm Thanks for the review, I will take a closer look over the next days...

gterzian commented on components/script/dom/htmlimageelement.rs:
```diff
@@ -271,106 +268,360 @@ impl HTMLImageElement {
+        match self.select_image_source() {
+            Some(src) => {
+                // Step 10
+                self.dispatch_progress_event(atom!("loadstart"), 0, None);
```
`dispatch_progress_event` dispatches a progress event via the scheduling of a task, is that what you mean? Should I change the method name to better reflect this?

jdm replied:
You're thinking of `dispatch_event`; `dispatch_progress_event` synchronously dispatches an event.

gterzian commented on components/script/dom/htmlimageelement.rs:
```diff
@@ -271,106 +268,360 @@ impl HTMLImageElement {
+                        self.abort_request(State::Broken, ImageRequestPhase::Pending);
+                        self.set_current_request_url_to_selected(src);
+                        self.dispatch_event(atom!("error"));
+                        self.dispatch_event(atom!("loadend"));
```
`dispatch_event` also does the dispatching by scheduling a task to do so, I think that's what you mean? Same as with https://github.com/servo/servo/pull/16238/files#r114092939

jdm replied:
According to the spec, setting the current request url and dispatching the events should happen as part of the same task. This is observable if the event handlers for those events uses microtasks, which would be executed in between those separate tasks in the current implementation.

gterzian commented:
@jdm thanks for the additional info, I've now:

1. Renamed `dispatch_progress_event` to `dispatch_loadstart_progress_event` and made it async. 
2. On a few other instances, made the various operations part of the same task, to match the spec.
3. Made an attempt to "verify that we only see the second src value that was set".

Please take another look when you have a chance.

jdm commented on components/script/dom/htmlimageelement.rs:
```diff
@@ -428,23 +411,55 @@ impl HTMLImageElement {
+                let mut current_request = img.current_request.borrow_mut();
+                current_request.source_url = Some(DOMString::from_string(self.src));
+                img.upcast::<EventTarget>().fire_event(atom!("error"));
+                img.upcast::<EventTarget>().fire_event(atom!("loadend"));
```
Let's do:
```rust
{
    let mut current_request = img.current_request.borrow_mut();
    current_request.source_url = Some(DOMString::from_string(self.src));
}
img.upcast::<EventTarget>().fire_event(atom!("error"));
img.upcast::<EventTarget>().fire_event(atom!("loadend"));
```
Since the events could call back into HTMLImageElement code, we run the risk of dynamic borrow failures if we leave the borrow unscoped.

jdm commented on components/script/dom/htmlimageelement.rs:
```diff
@@ -428,23 +411,55 @@ impl HTMLImageElement {
+                current_request.parsed_url = Some(self.url.clone());
                 current_request.source_url = Some(self.src.into());
+                // TODO: restart animation, if set
+                img.upcast::<EventTarget>().fire_event(atom!("load"));
```
Same as earlier - we need to scope the borrow.

jdm commented on components/script/dom/htmlimageelement.rs:
```diff
@@ -428,23 +411,55 @@ impl HTMLImageElement {
+    fn set_current_request_url_to_string_and_fire_load(&self, src: DOMString, url: ServoUrl) {
+        struct SetUrlToStringTask {
+            img: Trusted<HTMLImageElement>,
+            src: String,
```
Why not store a DOMString, since we seem to be converting to and from String?

gterzian replied:
Storing a DOMString gives the error `the trait `core::marker::Send` is not implemented for `dom::bindings::str::DOMString`, also it seems DOMString has `!Send` implemented (I tried adding an empty `Send` implementation, which gave me a `conflicting implementation` error...

jdm replied:
Oh, right. Carry on, in that case.

jdm reviewed (changes requested):
This looks great!

jdm commented:
The travis failure is unrelated to your changes and should be fixed by rebasing against master.

gterzian commented:
@jdm thanks, I've addressed your comments and rebased...

jdm commented:
@bors-servo: r+
Woohoo!

bors-servo commented:
:pushpin: Commit 1dca886 has been approved by `jdm`

<!-- @bors-servo r=jdm 1dca886962b5815737635093a3b780b96229146e -->

bors-servo commented:
:hourglass: Testing commit 1dca886962b5815737635093a3b780b96229146e with merge 701f27f83b96158adff4304d2d2341507b289f29...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/3940)

jdm commented:
```
  ▶ FAIL [expected PASS] /2dcontext/compositing/canvas_compositing_globalcompositeoperation_001.htm
  └   → /2dcontext/compositing/canvas_compositing_globalcompositeoperation_001.htm 5eb8afff87220c75e9a497c9a53d51f4fe15724a
/2dcontext/compositing/canvas_compositing_globalcompositeoperation_001-ref.htm 327b8fb9858ecf39e44061ddfa53995f6f09d286
Testing 5eb8afff87220c75e9a497c9a53d51f4fe15724a == 327b8fb9858ecf39e44061ddfa53995f6f09d286

  ▶ FAIL [expected PASS] /2dcontext/drawing-images-to-the-canvas/drawimage_html_image_12.html
  └   → /2dcontext/drawing-images-to-the-canvas/drawimage_html_image_12.html 23b755622cc81841a23cc8431a6546dfb1401c66
/2dcontext/drawing-images-to-the-canvas/drawimage_html_image_12_ref.html 99fb6d8c647f971713910d68e6d4c68f6df58e47
Testing 23b755622cc81841a23cc8431a6546dfb1401c66 == 99fb6d8c647f971713910d68e6d4c68f6df58e47

  ▶ FAIL [expected PASS] /2dcontext/the-canvas-state/canvas_state_restore_001.htm
  └   → /2dcontext/the-canvas-state/canvas_state_restore_001.htm 8f56a805ac02a49b7380f8146ddc79cc20a21fc5
/2dcontext/the-canvas-state/canvas_state_restore_001-ref.htm 0cfc22716a67fc8d6268bdf9e1f9865e103037b2
Testing 8f56a805ac02a49b7380f8146ddc79cc20a21fc5 == 0cfc22716a67fc8d6268bdf9e1f9865e103037b2

  ▶ FAIL [expected PASS] /2dcontext/drawing-images-to-the-canvas/drawimage_html_image_2.html
  └   → /2dcontext/drawing-images-to-the-canvas/drawimage_html_image_2.html 23b755622cc81841a23cc8431a6546dfb1401c66
/2dcontext/drawing-images-to-the-canvas/drawimage_html_image_2_ref.html f17d9836c26d18c3ed6caab0a4fe99f88ef7dae8
Testing 23b755622cc81841a23cc8431a6546dfb1401c66 == f17d9836c26d18c3ed6caab0a4fe99f88ef7dae8

  ▶ FAIL [expected PASS] /2dcontext/drawing-images-to-the-canvas/drawimage_html_image_10.html
  └   → /2dcontext/drawing-images-to-the-canvas/drawimage_html_image_10.html 23b755622cc81841a23cc8431a6546dfb1401c66
/2dcontext/drawing-images-to-the-canvas/drawimage_html_image_10_ref.html 4b2d047205d6594381d898d1183b6ef3bde1d343
Testing 23b755622cc81841a23cc8431a6546dfb1401c66 == 4b2d047205d6594381d898d1183b6ef3bde1d343

  ▶ FAIL [expected PASS] /2dcontext/drawing-images-to-the-canvas/drawimage_html_image_6.html
  └   → /2dcontext/drawing-images-to-the-canvas/drawimage_html_image_6.html 23b755622cc81841a23cc8431a6546dfb1401c66
/2dcontext/drawing-images-to-the-canvas/drawimage_html_image_6_ref.html b7dfcb2a815162af8df8566dd7661518e44d9e2d
Testing 23b755622cc81841a23cc8431a6546dfb1401c66 == b7dfcb2a815162af8df8566dd7661518e44d9e2d

  ▶ FAIL [expected PASS] /2dcontext/drawing-images-to-the-canvas/drawimage_html_image_7.html
  └   → /2dcontext/drawing-images-to-the-canvas/drawimage_html_image_7.html 23b755622cc81841a23cc8431a6546dfb1401c66
/2dcontext/drawing-images-to-the-canvas/drawimage_html_image_7_ref.html 5dab8f06767a0561d88ff99fcf619b55deaf5ee0
Testing 23b755622cc81841a23cc8431a6546dfb1401c66 == 5dab8f06767a0561d88ff99fcf619b55deaf5ee0

  ▶ FAIL [expected PASS] /2dcontext/drawing-images-to-the-canvas/drawimage_html_image_4.html
  └   → /2dcontext/drawing-images-to-the-canvas/drawimage_html_image_4.html cf89830df3610be45bd231cd96c08b22f8112d56
/2dcontext/drawing-images-to-the-canvas/drawimage_html_image_4_ref.html 38753d5bc881a3865835a6f04b6245fcf26b2ad3
Testing cf89830df3610be45bd231cd96c08b22f8112d56 == 38753d5bc881a3865835a6f04b6245fcf26b2ad3

  ▶ FAIL [expected PASS] /2dcontext/drawing-images-to-the-canvas/drawimage_html_image_8.html
  └   → /2dcontext/drawing-images-to-the-canvas/drawimage_html_image_8.html 23b755622cc81841a23cc8431a6546dfb1401c66
/2dcontext/drawing-images-to-the-canvas/drawimage_html_image_8_ref.html b7dfcb2a815162af8df8566dd7661518e44d9e2d
Testing 23b755622cc81841a23cc8431a6546dfb1401c66 == b7dfcb2a815162af8df8566dd7661518e44d9e2d

  ▶ FAIL [expected PASS] /2dcontext/drawing-images-to-the-canvas/drawimage_html_image_5.html
  └   → /2dcontext/drawing-images-to-the-canvas/drawimage_html_image_5.html d622872cfe463f511f76c01112a4bd305cb78ad3
/2dcontext/drawing-images-to-the-canvas/drawimage_html_image_5_ref.html 38753d5bc881a3865835a6f04b6245fcf26b2ad3
Testing d622872cfe463f511f76c01112a4bd305cb78ad3 == 38753d5bc881a3865835a6f04b6245fcf26b2ad3

  ▶ FAIL [expected PASS] /html/rendering/replaced-elements/attributes-for-embedded-content-and-images/img-dim.html
  └   → /html/rendering/replaced-elements/attributes-for-embedded-content-and-images/img-dim.html 23b755622cc81841a23cc8431a6546dfb1401c66
/html/rendering/replaced-elements/attributes-for-embedded-content-and-images/img-dim-ref.html a06da02484be571166d46ce1e1d5c6ed38bbd46f
Testing 23b755622cc81841a23cc8431a6546dfb1401c66 == a06da02484be571166d46ce1e1d5c6ed38bbd46f

  ▶ FAIL [expected PASS] /html/semantics/embedded-content/the-img-element/document-base-url.html
  └   → /html/semantics/embedded-content/the-img-element/document-base-url.html 23b755622cc81841a23cc8431a6546dfb1401c66
/html/semantics/embedded-content/the-img-element/document-base-url-ref.html e03c167dc3a5e3a0c744d05f921f4e3ff0b9ecd5
Testing 23b755622cc81841a23cc8431a6546dfb1401c66 == e03c167dc3a5e3a0c744d05f921f4e3ff0b9ecd5

  ▶ FAIL [expected PASS] /_mozilla/css/acid1_a.html
  └   → /_mozilla/css/acid1_a.html ba47affcf51260590a32f4043f119b20eb43040c
/_mozilla/css/acid1_b.html 47f030a74c6092a0ae8694d311af15c53dfae61d
Testing ba47affcf51260590a32f4043f119b20eb43040c == 47f030a74c6092a0ae8694d311af15c53dfae61d

  ▶ FAIL [expected PASS] /_mozilla/css/alpha_gif_a.html
  └   → /_mozilla/css/alpha_gif_a.html c89bb7a73bace01db22f61e74354fd9c720fc3df
/_mozilla/css/alpha_gif_b.html 45581bd3a4d5512803760844784ee5ea0784a0b4
Testing c89bb7a73bace01db22f61e74354fd9c720fc3df == 45581bd3a4d5512803760844784ee5ea0784a0b4

  ▶ FAIL [expected PASS] /_mozilla/css/background_size_a.html
  └   → /_mozilla/css/background_size_a.html adeb16477e6f837ec4193e0d4d448f98aa212809
/_mozilla/css/background_size_ref.html 23b755622cc81841a23cc8431a6546dfb1401c66
Testing adeb16477e6f837ec4193e0d4d448f98aa212809 == 23b755622cc81841a23cc8431a6546dfb1401c66

  ▶ FAIL [expected PASS] /_mozilla/css/block_replaced_content_b.html
  └   → /_mozilla/css/block_replaced_content_b.html d9bde2a48802b55baf0bfee9ae0d7a48190b1722
/_mozilla/css/block_replaced_content_ref.html 4e6ec2a4e7504e5c4bb739eb4fad793251d3180f
Testing d9bde2a48802b55baf0bfee9ae0d7a48190b1722 == 4e6ec2a4e7504e5c4bb739eb4fad793251d3180f

  ▶ FAIL [expected PASS] /_mozilla/css/canvas_radial_gradient_a.html
  └   → /_mozilla/css/canvas_radial_gradient_a.html bd0ad8026724b09481bf0c73b0c75f2253ccb754
/_mozilla/css/canvas_radial_gradient_ref.html 23b755622cc81841a23cc8431a6546dfb1401c66
Testing bd0ad8026724b09481bf0c73b0c75f2253ccb754 == 23b755622cc81841a23cc8431a6546dfb1401c66

  ▶ FAIL [expected PASS] /_mozilla/css/filter_inline_a.html
  └   → /_mozilla/css/filter_inline_a.html 40766601635fe01016a245b3ad17ac4ea0d3c4dd
/_mozilla/css/filter_inline_ref.html 23b755622cc81841a23cc8431a6546dfb1401c66
Testing 40766601635fe01016a245b3ad17ac4ea0d3c4dd == 23b755622cc81841a23cc8431a6546dfb1401c66

  ▶ FAIL [expected PASS] /_mozilla/css/image_rendering_pixelated_a.html
  └   → /_mozilla/css/image_rendering_pixelated_a.html 23b755622cc81841a23cc8431a6546dfb1401c66
/_mozilla/css/image_rendering_pixelated_ref.html 19a71361f3cc516651f155dddc0fcb9bd0e82fe9
Testing 23b755622cc81841a23cc8431a6546dfb1401c66 == 19a71361f3cc516651f155dddc0fcb9bd0e82fe9

  ▶ FAIL [expected PASS] /_mozilla/css/img_block_maxwidth_a.html
  └   → /_mozilla/css/img_block_maxwidth_a.html 6322442b705a283a7228e569f0aa2d8621332004
/_mozilla/css/img_block_maxwidth_ref.html 23b755622cc81841a23cc8431a6546dfb1401c66
Testing 6322442b705a283a7228e569f0aa2d8621332004 == 23b755622cc81841a23cc8431a6546dfb1401c66

  ▶ FAIL [expected PASS] /_mozilla/css/img_block_display_a.html
  └   → /_mozilla/css/img_block_display_a.html 2e52825aab60ff0bc02a285356ec786281191081
/_mozilla/css/img_block_display_ref.html 22c3558562e6cee5abcb068d566aae78bb8bb36a
Testing 2e52825aab60ff0bc02a285356ec786281191081 == 22c3558562e6cee5abcb068d566aae78bb8bb36a

  ▶ FAIL [expected PASS] /_mozilla/css/img_size_a.html
  └   → /_mozilla/css/img_size_a.html 23b755622cc81841a23cc8431a6546dfb1401c66
/_mozilla/css/img_size_b.html 9ba19abe46d51bcb7b5d19872c9dcb3149965935
Testing 23b755622cc81841a23cc8431a6546dfb1401c66 == 9ba19abe46d51bcb7b5d19872c9dcb3149965935

  ▶ FAIL [expected PASS] /_mozilla/css/img_width_style_intrinsic_width_a.html
  └   → /_mozilla/css/img_width_style_intrinsic_width_a.html d38e0056db72b48a67d22c85460ea2c57f3ed4c2
/_mozilla/css/img_width_style_intrinsic_width_ref.html d748c3d0718af645818bc66992b3c6ec439d57b4
Testing d38e0056db72b48a67d22c85460ea2c57f3ed4c2 == d748c3d0718af645818bc66992b3c6ec439d57b4

  ▶ FAIL [expected PASS] /_mozilla/css/inline_absolute_hypothetical_line_metrics_a.html
  └   → /_mozilla/css/inline_absolute_hypothetical_line_metrics_a.html 8a2cef34335fc6b40cce6a4a961c75fa44571a2f
/_mozilla/css/inline_absolute_hypothetical_line_metrics_ref.html 46fab4c4f4b9a9ed65af92a5b26aa7ca1f119c89
Testing 8a2cef34335fc6b40cce6a4a961c75fa44571a2f == 46fab4c4f4b9a9ed65af92a5b26aa7ca1f119c89

  ▶ FAIL [expected PASS] /_mozilla/css/inline_block_img_a.html
  └   → /_mozilla/css/inline_block_img_a.html b537ef7d510c2ffaa9b8014c8a32e71c125dc821
/_mozilla/css/inline_block_img_ref.html 127f17139f35047fb2073a8b6db180b39a2d9f5e
Testing b537ef7d510c2ffaa9b8014c8a32e71c125dc821 == 127f17139f35047fb2073a8b6db180b39a2d9f5e

  ▶ FAIL [expected PASS] /_mozilla/css/inline_margin_multiple_fragments_a.html
  └   → /_mozilla/css/inline_margin_multiple_fragments_a.html 23b755622cc81841a23cc8431a6546dfb1401c66
/_mozilla/css/inline_margin_multiple_fragments_ref.html 9f302380f7dfbfb148145a6f2fbad4c604ef2c06
Testing 23b755622cc81841a23cc8431a6546dfb1401c66 == 9f302380f7dfbfb148145a6f2fbad4c604ef2c06

  ▶ FAIL [expected PASS] /_mozilla/css/inline_stacking_context.html
  └   → /_mozilla/css/inline_stacking_context.html e5215fece263bc8309c25d5a7dd5c262bb7f515c
/_mozilla/css/inline_stacking_context_ref.html b84b80113bc7354c2ef8073834e00b5dc71722c6
Testing e5215fece263bc8309c25d5a7dd5c262bb7f515c == b84b80113bc7354c2ef8073834e00b5dc71722c6

  ▶ FAIL [expected PASS] /_mozilla/css/jpeg_progressive.html
  └   → /_mozilla/css/jpeg_progressive.html 23b755622cc81841a23cc8431a6546dfb1401c66
/_mozilla/css/jpeg_ref.html c4ab5226c96233ef33a09809c1e757748f4a460f
Testing 23b755622cc81841a23cc8431a6546dfb1401c66 == c4ab5226c96233ef33a09809c1e757748f4a460f

  ▶ FAIL [expected PASS] /_mozilla/css/list_style_position_a.html
  └   → /_mozilla/css/list_style_position_a.html 3a22316f0d9c96cd73bf2dc7e203ef7e29c0a030
/_mozilla/css/list_style_position_ref.html 23b755622cc81841a23cc8431a6546dfb1401c66
Testing 3a22316f0d9c96cd73bf2dc7e203ef7e29c0a030 == 23b755622cc81841a23cc8431a6546dfb1401c66

  ▶ FAIL [expected PASS] /_mozilla/css/object_element_a.html
  └   → /_mozilla/css/object_element_a.html d9c81c03ba77b4cf7d6960efeaa0119f705756a0
/_mozilla/css/object_element_b.html 23b755622cc81841a23cc8431a6546dfb1401c66
Testing d9c81c03ba77b4cf7d6960efeaa0119f705756a0 == 23b755622cc81841a23cc8431a6546dfb1401c66

  ▶ FAIL [expected PASS] /_mozilla/css/outset.html
  └   → /_mozilla/css/outset.html a5457764929399e7308960b8877c7ee86cda47a8
/_mozilla/css/outset_ref.html 23b755622cc81841a23cc8431a6546dfb1401c66
Testing a5457764929399e7308960b8877c7ee86cda47a8 == 23b755622cc81841a23cc8431a6546dfb1401c66

  ▶ FAIL [expected PASS] /_mozilla/css/png_rgba_colorspace_a.html
  └   → /_mozilla/css/png_rgba_colorspace_a.html 23b755622cc81841a23cc8431a6546dfb1401c66
/_mozilla/css/png_rgba_colorspace_b.html 2d5103c1d79069b6628ffe931d05456e5b9244b2
Testing 23b755622cc81841a23cc8431a6546dfb1401c66 == 2d5103c1d79069b6628ffe931d05456e5b9244b2

  ▶ FAIL [expected PASS] /_mozilla/css/position_abs_replaced_simple_a.html
  └   → /_mozilla/css/position_abs_replaced_simple_a.html a9405ea0f9f2fa74696ae2d7f14508f0a3325e1f
/_mozilla/css/position_abs_replaced_simple_b.html bcc7b6debd3d756df4131b07c0eeba6bdd6b1c27
Testing a9405ea0f9f2fa74696ae2d7f14508f0a3325e1f == bcc7b6debd3d756df4131b07c0eeba6bdd6b1c27

  ▶ FAIL [expected PASS] /_mozilla/css/vertical_align_middle_a.html
  └   → /_mozilla/css/vertical_align_middle_a.html 23b755622cc81841a23cc8431a6546dfb1401c66
/_mozilla/css/vertical_align_middle_ref.html bcc2be59add9b585e8d9cf282be376b3ac4a3cfd
Testing 23b755622cc81841a23cc8431a6546dfb1401c66 == bcc2be59add9b585e8d9cf282be376b3ac4a3cfd

  ▶ FAIL [expected PASS] /_mozilla/mozilla/FileAPI/blob_url_upload.html
  └   → /_mozilla/mozilla/FileAPI/blob_url_upload.html 7c3c0e7f0eb89a6c15142212544b7a9e8c331bb2
/_mozilla/mozilla/FileAPI/blob_url_upload_ref.html f860a8aaf24b5c2af599905b021a4d8a01df7afc
Testing 7c3c0e7f0eb89a6c15142212544b7a9e8c331bb2 == f860a8aaf24b5c2af599905b021a4d8a01df7afc

  ▶ FAIL [expected PASS] /_mozilla/mozilla/webgl/tex_image_2d_mipmap.html
  └   → /_mozilla/mozilla/webgl/tex_image_2d_mipmap.html 06711e6d8ba9520a7c0476d41130022d3e609160
/_mozilla/mozilla/webgl/tex_image_2d_simple_ref.html 23b755622cc81841a23cc8431a6546dfb1401c66
Testing 06711e6d8ba9520a7c0476d41130022d3e609160 == 23b755622cc81841a23cc8431a6546dfb1401c66

  ▶ Unexpected subtest result in /2dcontext/compositing/2d.composite.canvas.copy.html:
  │ FAIL [expected PASS] Canvas test: 2d.composite.canvas.copy
  └   → The object is in an invalid state.

  ▶ Unexpected subtest result in /2dcontext/compositing/2d.composite.canvas.destination-in.html:
  │ FAIL [expected PASS] Canvas test: 2d.composite.canvas.destination-in
  └   → The object is in an invalid state.

  ▶ Unexpected subtest result in /2dcontext/compositing/2d.composite.canvas.lighter.html:
  │ FAIL [expected PASS] Canvas test: 2d.composite.canvas.lighter
  └   → The object is in an invalid state.

  ▶ Unexpected subtest result in /2dcontext/compositing/2d.composite.canvas.source-in.html:
  │ FAIL [expected PASS] Canvas test: 2d.composite.canvas.source-in
  └   → The object is in an invalid state.

  ▶ Unexpected subtest result in /2dcontext/compositing/2d.composite.canvas.xor.html:
  │ FAIL [expected PASS] Canvas test: 2d.composite.canvas.xor
  └   → The object is in an invalid state.

  ▶ Unexpected subtest result in /2dcontext/compositing/2d.composite.image.copy.html:
  │ FAIL [expected PASS] Canvas test: 2d.composite.image.copy
  └   → The object is in an invalid state.

  ▶ Unexpected subtest result in /2dcontext/compositing/2d.composite.image.source-atop.html:
  │ FAIL [expected PASS] Canvas test: 2d.composite.image.source-atop
  └   → The object is in an invalid state.

  ▶ Unexpected subtest result in /2dcontext/compositing/2d.composite.uncovered.image.source-out.html:
  │ FAIL [expected PASS] drawImage() draws pixels not covered by the source object as (0,0,0,0), and does not leave the pixels unchanged.
  └   → The object is in an invalid state.

  ▶ Unexpected subtest result in /2dcontext/compositing/2d.composite.uncovered.image.destination-in.html:
  │ FAIL [expected PASS] drawImage() draws pixels not covered by the source object as (0,0,0,0), and does not leave the pixels unchanged.
  └   → The object is in an invalid state.

  ▶ Unexpected subtest result in /2dcontext/compositing/2d.composite.uncovered.pattern.destination-atop.html:
  │ FAIL [expected PASS] Pattern fill() draws pixels not covered by the source object as (0,0,0,0), and does not leave the pixels unchanged.
  └   → The object is in an invalid state.

  ▶ Unexpected subtest result in /2dcontext/conformance-requirements/2d.voidreturn.html:
  │ FAIL [expected PASS] void methods return undefined
  └   → The object is in an invalid state.

  ▶ Unexpected subtest result in /2dcontext/drawing-images-to-the-canvas/2d.drawImage.3arg.html:
  │ FAIL [expected PASS] Canvas test: 2d.drawImage.3arg
  └   → The object is in an invalid state.

  ▶ Unexpected subtest result in /2dcontext/compositing/2d.composite.uncovered.pattern.destination-in.html:
  │ FAIL [expected PASS] Pattern fill() draws pixels not covered by the source object as (0,0,0,0), and does not leave the pixels unchanged.
  └   → The object is in an invalid state.

  ▶ Unexpected subtest result in /2dcontext/drawing-images-to-the-canvas/2d.drawImage.animated.poster.html:
  │ FAIL [expected PASS] drawImage() of an APNG draws the poster frame
  └   → The object is in an invalid state.

  ▶ Unexpected subtest result in /2dcontext/drawing-images-to-the-canvas/2d.drawImage.9arg.sourcepos.html:
  │ FAIL [expected PASS] Canvas test: 2d.drawImage.9arg.sourcepos
  └   → The object is in an invalid state.

  ▶ Unexpected subtest result in /2dcontext/drawing-images-to-the-canvas/2d.drawImage.alpha.html:
  │ FAIL [expected PASS] Canvas test: 2d.drawImage.alpha
  └   → The object is in an invalid state.

  ▶ Unexpected subtest result in /2dcontext/drawing-images-to-the-canvas/2d.drawImage.negativedest.html:
  │ FAIL [expected PASS] Negative destination width/height represents the correct rectangle
  └   → The object is in an invalid state.

  ▶ Unexpected subtest result in /2dcontext/drawing-images-to-the-canvas/2d.drawImage.nowrap.html:
  │ FAIL [expected PASS] Stretched images do not get pixels wrapping around the edges
  └   → The object is in an invalid state.

  ▶ Unexpected subtest result in /2dcontext/fill-and-stroke-styles/2d.pattern.basic.type.html:
  │ FAIL [expected PASS] Canvas test: 2d.pattern.basic.type
  └   → The object is in an invalid state.

  ▶ Unexpected subtest result in /2dcontext/fill-and-stroke-styles/2d.pattern.modify.image1.html:
  │ FAIL [expected PASS] Canvas test: 2d.pattern.modify.image1
  └   → The object is in an invalid state.

  ▶ Unexpected subtest result in /2dcontext/fill-and-stroke-styles/2d.pattern.modify.image2.html:
  │ FAIL [expected PASS] Canvas test: 2d.pattern.modify.image2
  └   → The object is in an invalid state.

  ▶ Unexpected subtest result in /2dcontext/fill-and-stroke-styles/2d.pattern.paint.norepeat.coord1.html:
  │ FAIL [expected PASS] Canvas test: 2d.pattern.paint.norepeat.coord1
  └   → The object is in an invalid state.

  ▶ Unexpected subtest result in /2dcontext/fill-and-stroke-styles/2d.pattern.paint.orientation.image.html:
  │ FAIL [expected PASS] Image patterns do not get flipped when painted
  └   → The object is in an invalid state.

  ▶ Unexpected subtest result in /2dcontext/fill-and-stroke-styles/2d.pattern.paint.repeat.coord3.html:
  │ FAIL [expected PASS] Canvas test: 2d.pattern.paint.repeat.coord3
  └   → The object is in an invalid state.

  ▶ Unexpected subtest result in /2dcontext/fill-and-stroke-styles/2d.pattern.paint.norepeat.outside.html:
  │ FAIL [expected PASS] Canvas test: 2d.pattern.paint.norepeat.outside
  └   → The object is in an invalid state.

  ▶ Unexpected subtest result in /2dcontext/fill-and-stroke-styles/2d.pattern.paint.repeaty.basic.html:
  │ FAIL [expected PASS] Canvas test: 2d.pattern.paint.repeaty.basic
  └   → The object is in an invalid state.

  ▶ Unexpected subtest result in /2dcontext/fill-and-stroke-styles/2d.pattern.paint.repeat.coord1.html:
  │ FAIL [expected PASS] Canvas test: 2d.pattern.paint.repeat.coord1
  └   → The object is in an invalid state.

  ▶ Unexpected subtest result in /2dcontext/fill-and-stroke-styles/2d.pattern.paint.repeat.basic.html:
  │ FAIL [expected PASS] Canvas test: 2d.pattern.paint.repeat.basic
  └   → The object is in an invalid state.

  ▶ Unexpected subtest result in /2dcontext/fill-and-stroke-styles/2d.pattern.paint.repeat.outside.html:
  │ FAIL [expected PASS] Canvas test: 2d.pattern.paint.repeat.outside
  └   → The object is in an invalid state.

  ▶ Unexpected subtest result in /2dcontext/fill-and-stroke-styles/2d.pattern.paint.repeatx.basic.html:
  │ FAIL [expected PASS] Canvas test: 2d.pattern.paint.repeatx.basic
  └   → The object is in an invalid state.

  ▶ Unexpected subtest result in /2dcontext/shadows/2d.shadow.image.alpha.html:
  │ FAIL [expected PASS] Shadows are drawn correctly for partially-transparent images
  └   → The object is in an invalid state.

  ▶ Unexpected subtest result in /2dcontext/shadows/2d.shadow.pattern.transparent.1.html:
  │ FAIL [expected PASS] Shadows are not drawn for transparent fill patterns
  └   → The object is in an invalid state.

  ▶ Unexpected subtest result in /2dcontext/shadows/2d.shadow.image.basic.html:
  │ FAIL [expected PASS] Shadows are drawn for images
  └   → The object is in an invalid state.

  ▶ Unexpected subtest result in /2dcontext/shadows/2d.shadow.pattern.alpha.html:
  │ FAIL [expected PASS] Shadows are drawn correctly for partially-transparent fill patterns
  └   → The object is in an invalid state.

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-canvas-element/security.drawImage.image.html:
  │ FAIL [expected PASS] drawImage of different-origin image makes the canvas origin-unclean
  └   → The object is in an invalid state.

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-canvas-element/security.drawImage.canvas.html:
  │ FAIL [expected PASS] drawImage of unclean canvas makes the canvas origin-unclean
  └   → The object is in an invalid state.

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-canvas-element/security.pattern.canvas.fillStyle.html:
  │ FAIL [expected PASS] Setting fillStyle to a pattern of an unclean canvas makes the canvas origin-unclean
  └   → The object is in an invalid state.

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-canvas-element/security.pattern.canvas.timing.html:
  │ FAIL [expected PASS] Pattern safety depends on whether the source was origin-clean, not on whether it still is clean
  └   → The object is in an invalid state.

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-canvas-element/security.pattern.canvas.strokeStyle.html:
  │ FAIL [expected PASS] Setting strokeStyle to a pattern of an unclean canvas makes the canvas origin-unclean
  └   → The object is in an invalid state.

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-canvas-element/security.pattern.image.fillStyle.html:
  │ FAIL [expected PASS] Setting fillStyle to a pattern of a different-origin image makes the canvas origin-unclean
  └   → The object is in an invalid state.

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-canvas-element/security.pattern.create.html:
  │ FAIL [expected PASS] Creating an unclean pattern does not make the canvas origin-unclean
  └   → The object is in an invalid state.

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-canvas-element/security.pattern.cross.html:
  │ FAIL [expected PASS] Using an unclean pattern makes the target canvas origin-unclean, not the pattern canvas
  └   → The object is in an invalid state.

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-canvas-element/security.reset.html:
  │ FAIL [expected PASS] Resetting the canvas state does not reset the origin-clean flag
  └   → The object is in an invalid state.

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-canvas-element/security.pattern.image.strokeStyle.html:
  │ FAIL [expected PASS] Setting strokeStyle to a pattern of a different-origin image makes the canvas origin-unclean
  └   → The object is in an invalid state.

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/current-pixel-density/basic.html:
  │ FAIL [expected PASS] <img src="/images/green-256x256.png" data-expect="256">
  │   → assert_equals: width expected 256 but got 0
  │ 
  │ onload/</<@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/current-pixel-density/basic.html:27:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  │ test@http://web-platform.test:8000/resources/testharness.js:497:9
  │ onload/<@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/current-pixel-density/basic.html:25:5
  └ onload@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/current-pixel-density/basic.html:24:3

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/delay-load-event.html:
  │ FAIL [expected PASS] Image element delays window's load event
  │   → assert_true: expected true got false
  │ 
  │ @http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/delay-load-event.html:14:13
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  └ Test.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:1446:17

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/invalid-src.html:
  │ FAIL [expected PASS] src=""
  │   → assert_false: loadend event should not fired expected false got true
  │ 
  │ @http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/invalid-src.html:34:13
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  │ Test.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:1446:17
  │ Test.prototype.step_timeout/<@http://web-platform.test:8000/resources/testharness.js:1464:20
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  └ Test.prototype.step_func/<@http://web-platform.test:8000/resources/testharness.js:1430:20

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/update-the-image-data/fail-to-resolve.html:
  │ FAIL [expected PASS] <img src="//[">
  │   → assert_equals: expected "//[" but got ""
  │ 
  │ onload/</<@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/update-the-image-data/fail-to-resolve.html:20:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  │ test@http://web-platform.test:8000/resources/testharness.js:497:9
  │ onload/<@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/update-the-image-data/fail-to-resolve.html:19:5
  └ onload@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/update-the-image-data/fail-to-resolve.html:18:3
```

gterzian commented:
Looks like there is a problem with delaying the load event(and maybe other things),  I'll look into it...

gterzian commented:
@jdm Ok, I was actually only running the one `img.complete` test locally, which I guess wasn't a good idea LOL. 

Just from looking at the `invalid-src` test, I can tell the problem is I am first terminating the load blockers, and then firing the error event, making the test fail. So I guess it requires some clean up here and there(I guess always first firing the events, then terminating the loadblockers), and I am hoping that lots of tests are failing because of a general problem with delaying the "load" event(I'm thinking of all those "object is in an invalid state" test failures and others..)

gterzian commented:
well in any case, happy to have those tests to look at, I can see quite a lot of stuff relates directly to the algorithm, will go over those 😄

jdm commented:
Yeah, my suspicion here is that there are one or two small changes that will fix most of the failing tests.

gterzian commented:
@jdm Ok, thanks. Just from looking at some of those tests, I can in most cases see quite clearly what they relate to, so I'll add some prints here and there and start debugging...

gterzian commented:
@jdm Here is the problem I haven't been able to figure out:

- It seems that we need to prevent the load event, while awaiting a stable state.

Here is what I tried so far:
1. Simple add a load blocker to the current request at the beginning of the algorithm, before the stable state is reached. It seems in that case that the loadblocker is dropped at the end of function, after `ScriptThread::await_stable_state(task)`. 
2. I tried some hacks with adding a blocking load to the DocumentLoader manually, without using LoadBlocker(also tried using `Documentoader.events_inhibited` directly). The problem with that is that one test fails "document adopt base url", because it involves switching the document of the image. So in that case the document that is blocked before a stable state, isn't the same that is unblocked afterwards, so you get panics related to the document still having a blocking load, or being stuck in `events_inhibited`. 
3. I tried passing the actual document, or DocumentLoader, as part of the task that awaits a stable state, to keep track of the document we have blocking/unblocking and remedy the problems under 2, However it seems quite hard to do this in a way that is accepted by the compiler. 

I'm still looking into this, and if you have any ideas based on the above, they are welcome!

gterzian commented:
@jdm Ok the good news is that all the tests in "the-img-element" are now passing. The bad news is that I need some aspirine and it's extremely hacky at this point 😄 

I'm thinking of cleaning this up with some sort of `HashMap<ServoUrl, ImageRequest>`, where we would store a request with a blocker(or maybe just the blocker, but seems easier to store a request in a DOMRefCell), for every document with a different base url, and to unwind those blockers at the end. In fact the blockers of the current/pending request don't seem necessary or useful, since we only create those after the stable state has been reached, and that's too late to properly delay the load event and make all those tests pass. 

So what is actually necessary is a loadblocker to act as a kind of "transaction" from when the algorithm start until the end, and the algorithm can encounter several "document" which all need to be blocked separately (example: create an img in an iframe, then append it to the main document see https://github.com/servo/servo/blob/master/tests/wpt/web-platform-tests/html/semantics/embedded-content/the-img-element/document-adopt-base-url.html)

What do you think?

jdm commented:
It's interesting that the stable state is the issue here. We might be able to avoid that problem by fixing https://github.com/servo/servo/issues/15375, since microtasks should execute before the code that checks if documents are ready to load.

gterzian commented:
@jdm you're right, I experimented with https://github.com/servo/servo/issues/16861 and all tests are passing(I could also remove a timeout from the tests which I previously had added). So once we've fixed the microtask issue, I can rebase off master and the tests "should" be passing...

bors-servo commented:
:umbrella: The latest upstream changes (presumably #16913) made this pull request unmergeable. Please resolve the merge conflicts.

gterzian commented:
@jdm after switching to microtasks to await a stable state, the` img.complete` test are timing out. I've made a few observations:

1. It doesn't seem to relate to unterminated loadblockers. 
2. It doesn't seem to relate to a particular subtest either. Basically, if you remove either the `async src complete test` or if you remove the `imageInstance.src =` at line 55, the two tests that actually result in a fully loaded image, the test doesn't timeout. 
3. If you add the `long` meta, the test doesn't timeout either(actually it did timeout once even with the tag, on my machine). 
4. If you run the test independently, it doesn't timeout either. It only timeouts if you run all the `the-img-element` tests. 

 Since this wasn't a problem previously when the sync steps were run as part of a `Runnable`, I'm assuming it might relate to the fact that the Microtasks aren't actually multithreaded. At the same time, the initial waiting of a stable state and running of the sync steps don't seem to take that long, it seems that the responses just take longer than 10 secs if you have several one happening in the same test...

Do you have any idea what could cause those timeouts?

gterzian commented:
As a further update, I tried a few different things, such as running the "sync" steps in a task started right after stable state has been reached, and it didn't solve the timeout. On the other hand, when I accidentally bypassed the two `set_current_request_url_to_selected_fire_error_and_loadend` and `set_current_request_url_to_none_fire_error` methods, each of which schedules a task, the timeout disappeared...

This could suggest there are just too many tasks being scheduled, making the algorithm slow in some way...

gterzian commented:
@jdm could we try a full test run on the server? I'd like to see what others tests are timing out, and if they time out at all on the server...

jdm commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit 66861cc678d0365ff5098b6654f0a427fe3fe378 with merge 758f41f98fef229313d718177b1a157399f593a5...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/4247)

gterzian commented:
@jdm OK, fixed the double borrowing crash through scoping...

gterzian commented:
@jdm This one might be ready for another test run. The previous run I think confirmed that the timeouts where simply due to my machine being too slow, and the only code change that would require additional review is where the current request is set to Unavailable at the beginning of the algorithm, this allows an additional test to pass. The crashes from the previous run where due to the fact that I hadn't scoped that part, so the current_request was doubly borrowed(this wasn't apparent from running the img specific tests, since the images are not cached, and the double borrowing occured in the "we found a cached image" part).

jdm commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit 71b0c1016461ab51daa9fc13e6454be1c8553c97 with merge f4333fc1c2563469a2eb3bf5ac3306dfa5b37ceb...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/7212), [arm32](http://build.servo.org/builders/arm32/builds/7427), [arm64](http://build.servo.org/builders/arm64/builds/7464), [linux-dev](http://build.servo.org/builders/linux-dev/builds/7556), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/4312), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/4331), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/7535), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/20), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/19), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/4240), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/4305), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/20), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/21), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/2779)
State: approved= try=True

jdm commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit 71b0c10 has been approved by `jdm`

<!-- @bors-servo r=jdm 71b0c1016461ab51daa9fc13e6454be1c8553c97 -->

bors-servo commented:
:hourglass: Testing commit 71b0c1016461ab51daa9fc13e6454be1c8553c97 with merge 7f5bc7341241deae3b7d7a2bb1f4194dcc7385ee...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/4310)

jdm commented:
@bors-servo: retry
* #17096

bors-servo commented:
:hourglass: Testing commit 71b0c1016461ab51daa9fc13e6454be1c8553c97 with merge fe7d039416ea93f9a5a120cda9a6114ec1438c3e...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/7227), [arm32](http://build.servo.org/builders/arm32/builds/7443), [arm64](http://build.servo.org/builders/arm64/builds/7480), [linux-dev](http://build.servo.org/builders/linux-dev/builds/7571), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/4327), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/4347), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/7551), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/35), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/34), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/4255), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/4320), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/36), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/37), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/2795)
Approved by: jdm
Pushing fe7d039416ea93f9a5a120cda9a6114ec1438c3e to master...

gterzian commented:
@jdm that was a big one, thanks for the efforts you put into reviewing this!
